src/browse.ts

```diff
@@ -356,14 +372,64 @@ export async function summarize (page: Page, keywords: string[], action: string)
   return resp.replace(/\n+/g, '\n')
 }
 
+function hasNoNonTextChildren (elem: cheerio.Element): boolean {
+  for (const child of elem.children) {
+    if (child.nodeType !== 3) {
+      return false
+    }
+  }
+  return true
+}
+
 export async function getText (locator: Locator): Promise<string> {
   let result = ''
-  result += await locator.getAttribute('aria-label') ?? '' + ' '
-  result += await locator.getAttribute('label') ?? '' + ' '
-  result += await locator.getAttribute('href') ?? '' + ' '
-  result += await locator.getAttribute('placeholder') ?? '' + ' '
+  result += await locator.getAttribute('aria-label', { timeout: 3000 }) ?? '' + ' '
+  result += await locator.getAttribute('label', { timeout: 3000 }) ?? '' + ' '
+  result += await locator.getAttribute('href', { timeout: 3000 }) ?? '' + ' '
+  result += await locator.getAttribute('placeholder', { timeout: 3000 }) ?? '' + ' '
 
   result += await locator.innerText()
 
   return result
 }
+
+const findShadowRoot = (): string => {
+  const queue: Element[] = [document.body]
+  let html = ''
+
+  while (queue.length > 0) {
+    const elem = queue.shift()
+
+    if ((elem?.shadowRoot) !== null) {
+      html += elem?.shadowRoot.innerHTML // Shadow root found
+    }
+
+    // Add all child elements to the queue for BFS
+    if (elem?.children != null) {
+      queue.push(...Array.from(elem?.children))
+    }
+  }
+
+  return html
+}
+
+async function getPageHTML (page: Page): Promise<string> {
```

Comment:
Elements in shadow DOMs are not included in `page.content()`, so we need to search for those and attach them to the end.